site/content/entry/improvements-to-hardware-renderer.md

```diff
@@ -0,0 +1,200 @@
+
+High end GPUs benefit the most
+graph: performance of several high end GPUs
+
+Low end and integrated improved but just a little less
+graph: performance of low end gpus
+
+## Obstacles to Emulating the PICA200 on a GPU
+
+### Making Functions Out of GOTOs
+
+It's likely that the game developers for the 3DS didn't have to write PICA200 GPU assembly, but when emulating the PICA200, all Citra can work with is a commandlist and a stream of PICA200 opcodes.
+While the developers probably wrote in a high level shader language that supports functions, when the shaders are compiled, most of that goes away.
+The PICA200 supports barebones `CALL`, `IF`, and `LOOP` operations, but also supports an arbitrary `JMP` that can go to any address.
+Translating PICA200 shaders into GLSL (OpenGL Shader Language) means that you'll have to be prepared to rewrite every arbitrary `JMP` without using a `goto` as GLSL doesn't support them.
+
+phantom assumed the worst when they originally translated PICA200 shaders into GLSL and wrote a monsterous switch statement that would have a case for every jump target and act as a PICA200 shader interpreter.
+This worked, but proved to be slower than the software renderer!
+Now that phantom knew it was possible, and they had some data about how the average PICA200 shader looked, they took to rewrite it with the goal to make it fast.
+While the shaders could theoretically be very unruly and hard to convert, almost all the shaders were well behaved, presumably because they are compiled from a higher level language.
+This time around, phantom generated native GLSL functions wherever possible by analyizing the control flow of the instructions, and the results are much prettier and faster.
+Armed with the new knowledge, phantom rewrote the conversion a third time, and optimized the generated shaders even further.
+What started off slower than the software renderer ended up being the massive performance boost we have today!
+
+### Geometry Shaders Are Back With a Vengence
+
+The veteran Citra user will remember a time long ago when Citra didn't support Geometry Shaders.
+It was a dark time where [Fire Emblem was mostly unplayable](https://github.com/citra-emu/citra/issues/2548) and [Pokémon didn't have attack animations](https://github.com/citra-emu/citra/issues/2629).
+Geometry shaders is a GPU stage where the game can transform a single primitive into zero or more primitives.
+This can be used to create particle effects from a few points or can be used to fill in gaps in a 3D model.
+Citra has supported Geometry Shaders since the summer of 2016 thanks to the heroic efforts of the many people that have worked on it in the past.
+
+When trying to convert the PICA200 geometry shaders into GLSL, phantom ran into some very tough problems.
+Some games rely on the shader state carrying over from run to run, which isn't something that you can do in GLSL.
+Even more so, some games rely on (i know i'm forgetting something here??????).
```

Comment:
remove

Reply:
I really wanted to add a segment on geometry shaders, but i honestly don't know enough about the challenges of supporting them in GLSL so I'll just remove the entire paragraph. We can re add it if someone who knows a thing or two about them chimes in before we go live.